libkbfs/block_retrieval_queue.go

```diff
@@ -249,27 +305,42 @@ func (brq *blockRetrievalQueue) CacheAndPrefetch(ctx context.Context,
 			close(didUpdateCh)
 		}
 	}()
-	// TODO: If the working set cache has triggered a prefetch in the past, and
-	// we've switched to a sync cache, we still need to re-trigger prefetches
-	// so that it goes down the tree.
-	// In that case, this `if` should check for `finishedPrefetch` instead.
-	if triggeredPrefetch {
+	if prefetchStatus == FinishedPrefetch {
+		// Finished prefetches can always be short circuited and respond on the
+		// success channel (upstream prefetches might need to block on other
+		// parallel prefetches too).
+		prefetchDoneCh <- struct{}{}
+		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(),
+			block, lifetime, FinishedPrefetch)
+	}
+	if brq.config.IsSyncedTlf(kmd.TlfID()) && dbc != nil {
+		switch prefetchStatus {
+		case NoPrefetch:
+		case TriggeredPrefetch:
+			// We still need to register this requestor for receiving
+			// notifications about this prefetch.
+			// TODO: handle
```

Comment:
Actually I can remove this TODO. Since I gave you the commit range link, I'll leave it alone for now.

Reply:
Hmm, yeah. I used to have `FinishedPrefetch` in here, but yes the switch is redundant now. Do you think I should leave the `if` here with a comment, or invert the logic and `AND` it with the `else` block?

Reply:
Whatever you think the clearest comment structure is would be fine with me, though w/ a very slight preference to a single `if` statement if the comment is clear enough.  Now I'm confused about the `else` block though.  Why is it ok to signal an error for prefetches that are still in progress?  If the caller supplied a channel, wouldn't they prefer to learn when the prefetch completes?

Reply:
All callers supply channels via `CacheAndPrefetch`. The `else` block is for non-synced TLFs.

Reply:
Why would someone requesting a non-synced prefetch not want to know when the existing prefetch has finished?  Not sure what the TLF type has to do with it.

Reply:
Because the `prefetchDoneCh` is only to signal that the full tree has finished prefetching.

Reply:
Ok, guess that makes sense for now, though it might need to change for git when full-tree prefetches in an unsynced TLF will be possible.